src/color.cc

```diff
@@ -213,6 +213,47 @@ ColorModule::~ColorModule()
 {
 }
 
+// Parses hex colors according to: https://drafts.csswg.org/css-color/#typedef-hex-color.
+// If the input is invalid, returns boost::none.
+// Supports the following formats:
+// * "#rrggbb"
+// * "#rrggbbaa"
+// * "#rgb"
+// * "#rgba"
+boost::optional<Color4f> parse_hex_color(const std::string& hex) {
+	// validate size. short syntax uses one hex digit per color channel instead of 2.
+	const bool short_syntax = hex.size() == 4 || hex.size() == 5;
+	const bool long_syntax = hex.size() == 7 || hex.size() == 9;
+	if (!short_syntax && !long_syntax) return boost::none;
+
+	if (hex[0] != '#') return boost::none;
+
+	// number of characters per color channel
+	const int stride = short_syntax ? 1 : 2;
+	const float channel_max = short_syntax ? 15 : 255;
+
+	std::array<int,4> rgba;
+	rgba[3] = channel_max; // default alpha channel to 100%
+
+	for (unsigned i = 0; i < (hex.size() - 1) / stride; i++) {
+		const std::string chunk = hex.substr(1 + i*stride, stride);
+
+		// validate - only [0-9a-fA-F] are allowed
+		for (char c : chunk) {
+			if (!std::isdigit(c) && !('a' <= c && c <= 'f') && !('A' <= c && c <= 'F')) return boost::none;
```

Comment:
Could you use `isxdigit(c)` instead?

..or maybe even smht. like `if (!std::all_of(std::cbegin(chunk), std::cend(chunk), std::isxdigit)) return boost::none;`

Reply:
Thanks, I hadn't seen `isxdigit` before, that's a lot cleaner.

I switched to `std::all_of`, but unfortunately `isxdigit` requires a cast from `char`, so I don't think there's a way to use it there without wrapping it in a lambda (see "Notes" [here](http://en.cppreference.com/w/cpp/string/byte/isxdigit)). Let me know if you see a better way to do that.

Reply:
Makes sense. Readability is a primary goal, so let's keep it easy.